examples/fine-tuned-RAG/ModelFinetune.ipynb

```diff
@@ -0,0 +1,1186 @@
+       "  0%|          | 0/100 [00:00<?, ?it/s]"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "# Prepare the OpenAI File format i.e. JSONL from train_sample\n",
+    "def dataframe_to_jsonl(df):\n",
+    "    def create_jsonl_entry(row):\n",
+    "        messages = row[\"few_shot_prompt_1K\"]\n",
+    "        return json.dumps({\"messages\": messages})\n",
+    "\n",
+    "    jsonl_output = df.progress_apply(create_jsonl_entry, axis=1)\n",
+    "    return \"\\n\".join(jsonl_output)\n",
+    "\n",
+    "with open(\"local_cache/100_train_few_shot.jsonl\", \"w\") as f:\n",
+    "    f.write(dataframe_to_jsonl(train_sample))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 65,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "File Status:  uploaded\n",
+      "File Status:  uploaded\n",
+      "File Status:  processed\n",
+      "Job Status:  running\n",
+      "Job Status:  running\n",
+      "Job Status:  running\n",
+      "Job Status:  running\n",
+      "Job Status:  running\n",
+      "Job Status:  running\n",
+      "Job Status:  running\n",
+      "Job Status:  running\n",
+      "Job Status:  running\n",
+      "Job Status:  running\n",
+      "Job Status:  running\n",
+      "Job Status:  running\n",
+      "Job Status:  running\n",
+      "Job Status:  running\n",
+      "Job Status:  running\n",
+      "Job Status:  running\n",
+      "Job Status:  running\n",
+      "Job Status:  running\n",
+      "Job Status:  running\n",
+      "Job Status:  running\n",
+      "Job Status:  succeeded\n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "'ft:gpt-3.5-turbo-0613:qdrant:trnfewshot20230907:7w8Fhrg6'"
+      ]
+     },
+     "execution_count": 65,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "fine_tuner = OpenAIFineTuner(\n",
+    "        training_file_path=\"local_cache/100_train_few_shot.jsonl\",\n",
+    "        model_name=\"gpt-3.5-turbo\",\n",
+    "        suffix=\"trnfewshot20230907\"\n",
+    "    )\n",
+    "\n",
+    "model_id = fine_tuner.fine_tune_model()\n",
+    "model_id"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 66,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "application/vnd.jupyter.widget-view+json": {
+       "model_id": "418ad403b26e43eeba085657e063046e",
+       "version_major": 2,
+       "version_minor": 0
+      },
+      "text/plain": [
+       "  0%|          | 0/100 [00:00<?, ?it/s]"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "df[\"ft_generated_answer_few_shot\"] = df.progress_apply(answer_question, model=model_id, prompt_func=get_few_shot_prompt, axis=1)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 7. Comparison and Results"
```

Comment:
This section is similarly confusing - can you summarize the takeaways under each chart? Currently it looks like the fine-tuning is the weakest approach using the bottom chart, but using the second bottom its the best, which means there is some trade-off that you should spell out to the readers here. It may be that one approach works better in some scenarios etc, so delve into that a little.